naga/src/back/msl/writer.rs

```diff
@@ -5873,12 +5941,149 @@ template <typename A>
+        writeln!(self.out, "{l2}{params}.dst_tf.k * dstLinearRgb,")?;
+        writeln!(self.out, "{l2}dstLinearRgb < {params}.dst_tf.b);")?;
+
+        writeln!(self.out, "{l1}return float4(dstGammaRgb, 1.0);")?;
+        Ok(())
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn write_wrapped_image_load(
+        &mut self,
+        module: &crate::Module,
+        func_ctx: &back::FunctionCtx,
+        image: Handle<crate::Expression>,
+        _coordinate: Handle<crate::Expression>,
+        _array_index: Option<Handle<crate::Expression>>,
+        _sample: Option<Handle<crate::Expression>>,
+        _level: Option<Handle<crate::Expression>>,
+    ) -> BackendResult {
+        // We currently only need to wrap image loads for external textures
+        let class = match *func_ctx.resolve_type(image, &module.types) {
+            crate::TypeInner::Image { class, .. } => class,
+            _ => unreachable!(),
+        };
+        if class != crate::ImageClass::External {
+            return Ok(());
+        }
+        let wrapped = WrappedFunction::ImageLoad { class };
+        if !self.wrapped_functions.insert(wrapped) {
+            return Ok(());
+        }
+
+        writeln!(self.out, "float4 {IMAGE_LOAD_EXTERNAL_FUNCTION}({EXTERNAL_TEXTURE_WRAPPER_STRUCT} tex, uint2 coords) {{")?;
+        let l1 = back::Level(1);
+        let l2 = l1.next();
+        let l3 = l2.next();
+        writeln!(
+            self.out,
+            "{l1}uint2 plane0_size = uint2(tex.plane0.get_width(), tex.plane0.get_height());"
+        )?;
+        // Clamp coords to provided size of external texture to prevent OOB
+        // read. If params.size is zero then clamp to the actual size of the
+        // texture.
+        writeln!(
+            self.out,
+            "{l1}uint2 cropped_size = {NAMESPACE}::any(tex.params.size != 0) ? tex.params.size : plane0_size;"
+        )?;
+        writeln!(
+            self.out,
+            "{l1}coords = {NAMESPACE}::min(coords, cropped_size - 1);"
+        )?;
+
+        // Apply load transformation
+        writeln!(self.out, "{l1}uint2 plane0_coords = uint2({NAMESPACE}::round(tex.params.load_transform * float3(float2(coords), 1.0)));")?;
+        writeln!(self.out, "{l1}if (tex.params.num_planes == 1u) {{")?;
+        // For single plane, simply read from plane0
+        writeln!(self.out, "{l2}return tex.plane0.read(plane0_coords);")?;
+        writeln!(self.out, "{l1}}} else {{")?;
+
+        // Chroma planes may be subsampled so we must scale the coords accordingly.
+        writeln!(
+            self.out,
+            "{l2}uint2 plane1_size = uint2(tex.plane1.get_width(), tex.plane1.get_height());"
+        )?;
+        writeln!(self.out, "{l2}uint2 plane1_coords = uint2({NAMESPACE}::round(float2(plane0_coords) / float2(plane0_size - 1) * float2(plane1_size - 1)));")?;
```

Comment:
If the size of plane0 is 1, we will divide by 0 here.

The HLSL implementation uses floor instead: `floor(plane0_coords * plane1_size / plane0_size)`.

Reply:
ugh sorry I'd fixed that long ago in the HLSL backend but forgot to change the MSL one too.